File: src/Benchmark/Types/Response.hs
-- | Request endpoint and HTTP response data types.
module Benchmark.Types.Response
  ( -- * Request/Response
    Endpoint (..)
  , TestingResponse (..)

    -- * Validation
  , FieldAssertion (..)
  , ValidationSpec (..)
  , ValidationError (..)
  , ValidationSummary (..)
  )
where

import Benchmark.Types.Internal (dropFieldPrefix)
import Benchmark.Types.Units (Nanoseconds)
import Data.Aeson
  ( FromJSON (..)
  , ToJSON (..)
  , Value
  , defaultOptions
  , fieldLabelModifier
  , genericParseJSON
  , object
  , withObject
  , (.:)
  , (.:?)
  , (.=)
  )
import Data.ByteString.Lazy qualified as LBS
import Data.Foldable (asum)
import Data.Map.Strict (Map)
import Data.Maybe (catMaybes)
import Data.Scientific (Scientific)
import Data.Text (Text)
import Data.Time (UTCTime)
import GHC.Generics (Generic)

-- | Per-field assertion in a validation spec.
data FieldAssertion
  = -- | Assert the key exists (any value, including null)
    FieldPresent
  | -- | Assert exact value match
    FieldEq Value
  | -- | Assert field is explicitly @null@
    FieldNull
  | -- | Assert field exists and is not @null@
    FieldNotNull
  | -- | Assert JSON type: @"string"@, @"number"@, @"boolean"@, @"array"@, @"object"@, @"null"@
    FieldType Text
  | -- | Assert string field matches a regex pattern (POSIX ERE)
    FieldMatches Text
  | -- | Assert numeric field is within [min, max] (both bounds optional)
    FieldRange (Maybe Scientific) (Maybe Scientific)
  | -- | Assert array has exactly N elements
    ArrayLength Int
  | -- | Assert a value is present in an array
    ArrayContains Value
  deriving stock (Show, Eq, Generic)

instance FromJSON FieldAssertion where
  parseJSON = withObject "FieldAssertion" $ \o ->
    asum
      [ FieldPresent <$ (o .: "present" >>= \b -> if (b :: Bool) then pure () else fail "present must be true")
      , FieldEq <$> o .: "eq"
      , FieldNull <$ (o .: "null" >>= \b -> if (b :: Bool) then pure () else fail "null must be true")
      , FieldNotNull <$ (o .: "notNull" >>= \b -> if (b :: Bool) then pure () else fail "notNull must be true")
      , FieldType <$> o .: "type"
      , FieldMatches <$> o .: "matches"
      , o .: "range" >>= withObject "range" (\r -> FieldRange <$> r .:? "min" <*> r .:? "max")
      , ArrayLength <$> o .: "arrayLength"
      , ArrayContains <$> o .: "arrayContains"
      ]

instance ToJSON FieldAssertion where
  toJSON FieldPresent = object ["present" .= True]
  toJSON (FieldEq v) = object ["eq" .= v]
  toJSON FieldNull = object ["null" .= True]
  toJSON FieldNotNull = object ["notNull" .= True]
  toJSON (FieldType t) = object ["type" .= t]
  toJSON (FieldMatches pat) = object ["matches" .= pat]
  toJSON (FieldRange mmin mmax) =
    object ["range" .= object (catMaybes [("min" .=) <$> mmin, ("max" .=) <$> mmax])]
  toJSON (ArrayLength n) = object ["arrayLength" .= n]
  toJSON (ArrayContains v) = object ["arrayContains" .= v]

-- | Declarative validation rules applied to every response for an endpoint.
data ValidationSpec = ValidationSpec
  { validateStatus :: Maybe Int
  -- ^ Expected HTTP status code
  , validateFields :: Maybe (Map Text FieldAssertion)
  -- ^ Map from dot-path (e.g. "$.user.id") to assertion
  }
  deriving stock (Show, Eq, Generic)

instance FromJSON ValidationSpec where
  parseJSON =
    genericParseJSON
      defaultOptions
        { fieldLabelModifier = dropFieldPrefix "validate"
        }

instance ToJSON ValidationSpec where
  toJSON spec =
    object $
      maybe [] (\s -> ["status" .= s]) (validateStatus spec)
        ++ maybe [] (\f -> ["fields" .= f]) (validateFields spec)

-- | A single validation failure on a response.
data ValidationError
  = -- | Status code did not match: expected, actual
    StatusCodeMismatch Int Int
  | -- | Field path was not found in the response body
    FieldNotFound Text
  | -- | Field path found but value differed: path, expected, actual
    FieldValueMismatch Text Value Value
  | -- | Response body was absent (no body returned)
    BodyAbsent
  | -- | Response body was present but could not be decoded as JSON
    BodyInvalidJSON
  deriving stock (Show, Eq, Generic)
  deriving anyclass (ToJSON)

-- | Aggregate validation results for all responses from one endpoint.
data ValidationSummary = ValidationSummary
  { totalValidated :: Int
  , totalFailed :: Int
  , validationErrors :: [ValidationError]
  }
  deriving stock (Show, Eq, Generic)
  deriving anyclass (ToJSON)

-- | An HTTP endpoint to benchmark, with method, URL, optional body, headers, and validation.
data Endpoint = Endpoint
  { method :: Text
  , url :: Text
  , body :: Maybe Value
  , headers :: [(Text, Text)]
  , validate :: Maybe ValidationSpec
  -- ^ Optional per-response validation rules
  }
  deriving stock (Show, Eq)

-- | Result of a single benchmarked HTTP request.
data TestingResponse = TestingResponse
  { durationNs :: Nanoseconds
  , statusCode :: Int
  , respBody :: Maybe LBS.ByteString
  , errorMessage :: Maybe String
  , requestedAt :: UTCTime
  }
  deriving stock (Show, Eq)

File: crates/gauntlet-core/src/types/response.rs
//! Response-validation config and the runtime response record. [`FieldAssertion`]
//! is externally tagged: unit checks are bare strings, data checks single-key
//! objects.

use std::collections::BTreeMap;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use super::config::HttpMethod;
use super::units::{ns_to_ms, Nanoseconds};

/// A per-field assertion in a validation spec.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum FieldAssertion {
    /// Key exists (any value, including null).
    Present,
    /// Field is explicitly `null`.
    Null,
    /// Field exists and is not `null`.
    NotNull,
    /// Exact value match.
    Eq(serde_json::Value),
    /// JSON type: `"string"`, `"number"`, `"boolean"`, `"array"`, `"object"`, `"null"`.
    Type(String),
    /// String field matches a POSIX ERE.
    Matches(String),
    /// Numeric field within `[min, max]` (both bounds optional).
    Range {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        min: Option<f64>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        max: Option<f64>,
    },
    /// Array has exactly N elements.
    ArrayLength(usize),
    /// Value is present in an array.
    ArrayContains(serde_json::Value),
}

/// Declarative validation rules applied to every response for an endpoint.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
pub struct ValidationSpec {
    /// Expected HTTP status code.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<u16>,
    /// Map from dot-path (e.g. `$.user.id`) to assertion.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fields: Option<BTreeMap<String, FieldAssertion>>,
}

/// An HTTP endpoint to benchmark: method, URL, optional body, headers, and
/// optional validation. Produced by `build_endpoints`; a runtime type.
#[derive(Clone, Debug, PartialEq)]
pub struct Endpoint {
    pub method: HttpMethod,
    pub url: String,
    pub body: Option<serde_json::Value>,
    pub headers: Vec<(String, String)>,
    pub validate: Option<ValidationSpec>,
}

/// The result of a single benchmarked HTTP request, constructed by the engine.
/// A `Some(error)` marks a transport failure, excluded from latency math.
#[derive(Clone, Debug, PartialEq)]
pub struct TestingResponse {
    pub duration: Nanoseconds,
    pub status: u16,
    pub error: Option<String>,
}

impl TestingResponse {
    /// This response's latency in milliseconds, or `None` if the request failed.
    pub fn latency_ms(&self) -> Option<f64> {
        match self.error {
            Some(_) => None,
            None => Some(ns_to_ms(self.duration).0),
        }
    }
}

/// Extract the successful-response latencies (ms) from a batch, dropping
/// failures. The feed into `gauntlet_stats::calculate_stats`.
pub fn extract_durations(responses: &[TestingResponse]) -> Vec<f64> {
    responses
        .iter()
        .filter_map(TestingResponse::latency_ms)
        .collect()
}

/// A single failed assertion against a response: the dot-path that failed and a
/// human-readable reason. Produced by the engine's validation pass.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct ValidationError {
    /// The field path that failed (e.g. `$.user.id`), or `status` for the
    /// status-code check.
    pub field: String,
    /// Why it failed.
    pub message: String,
}

/// Aggregate validation outcome for one endpoint. Errors are capped at
/// [`MAX_VALIDATION_ERRORS`].
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct ValidationSummary {
    pub total: usize,
    pub failed: usize,
    pub errors: Vec<ValidationError>,
}

/// Cap on retained validation errors per endpoint, so a fully-failing run cannot
/// accumulate unbounded detail.
pub const MAX_VALIDATION_ERRORS: usize = 50;
